feat: allow users to add insights to multiple dashboards

pauldambra opened the pull request:
## Problem

stacked on top of #9524 this adds UI to allow testing addition of insights to multiple dashboards

Design at https://www.figma.com/file/gQBj9YnNgD8YW4nBwCVLZf/PostHog-App?node-id=8803%3A44524

## Changes

![multidash-examples](https://user-images.githubusercontent.com/984817/165336878-71b6dcc8-9a52-43ab-9373-62221823ffbd.gif)

Adds a new modal behind a feature flag that allows you to

* see which dashboards an insight is on
* search a list of dashboards
* remove insight from dashboards
* add insight to a dashboard
* create a new dashboard (so an insight can be added to it)

NB while this is safe to merge behind a flag. We likely need to add a way to have more than one `last_refresh` time for an insight before we can start using this widely

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

running it locally and using it to add and remove dashboards

pauldambra committed:
spike the feature

pauldambra committed:
generate tiles filters_hash if there isn't one on read

pauldambra committed:
don't filter a get by first

pauldambra committed:
Merge branch 'master' into multiple_filter_hashes

pauldambra committed:
en-nicen a test

pauldambra committed:
obey github's mypy

pauldambra committed:
obey github's mypy

pauldambra committed:
add feature flag

pauldambra committed:
Add gauge icon to button

pauldambra committed:
add counted icon

pauldambra committed:
handle zero better in counted icon

pauldambra committed:
show dashboard count button when ff is on

pauldambra committed:
merge from master

pauldambra committed:
merge from master

pauldambra committed:
add a new (incomplete) modal to replace the existing save to dashboard modal

pauldambra committed:
add search box and existing count text without functionality

pauldambra committed:
badly add an infinite list to the new modal

pauldambra committed:
slightly nicer layout

pauldambra committed:
make the list of dashboards ordered and searchable

pauldambra committed:
modal UI without functionality in place

pauldambra committed:
wire up close button

pauldambra committed:
allow adding to dashboard from the modal

pauldambra committed:
wire up remove from dashboard

pauldambra committed:
wire up add, remove, and remove all buttons

pauldambra committed:
wire up add to new dashboard button

pauldambra committed:
force type to be number not literal -1

pauldambra committed:
do not duplicate insights in results when querying for saved

pauldambra committed:
removes the 'remove all' clickable text

pauldambra committed:
Merge branch 'master' into multiple_filter_hashes

pauldambra committed:
Merge branch 'multiple_filter_hashes' into add_multi_dash_ui

pauldambra committed:
remove unnecessary Q

pauldambra committed:
more specific exception

pauldambra committed:
reduce queries by storing filters_hash in context

pauldambra commented:
Add to dashboard button

<img width="194" alt="Screenshot 2022-04-26 at 17 00 35" src="https://user-images.githubusercontent.com/984817/165343422-addd0806-092b-4a46-9cbf-46dee5a40e4b.png">

<img width="197" alt="Screenshot 2022-04-26 at 17 00 18" src="https://user-images.githubusercontent.com/984817/165343442-e32c6b07-fd4f-4d9f-b055-ca0c34aef959.png">

<img width="195" alt="Screenshot 2022-04-26 at 17 00 56" src="https://user-images.githubusercontent.com/984817/165343462-14c3f897-173f-468e-bd67-e25f46a0cdc6.png">

pauldambra commented:
add to dashboard modal

<img width="539" alt="Screenshot 2022-04-26 at 17 03 22" src="https://user-images.githubusercontent.com/984817/165343601-0fb2bc92-4b50-4a37-8081-99703fd9b7c3.png">

pauldambra commented:
tagging @lottiecoxon and @clarkus for review since the UI spans their handover :)

clarkus reviewed (changes requested):
Looks great. The one opportunity I see is to automatically add the insight to a newly created dashboard. I noticed in your gif that you created the dashboard, it appended to the list, you clicked to add the insight. Can we collapse all that into the submit action for adding a new dashboard? For example:

1. Click to create a new dashboard
2. Name the dashboard and submit
3. See the new dashboard added to the list and shown as `added` 

Thoughts?

pauldambra committed:
make filters_hash readonly and document why it is there

pauldambra committed:
remove console log on scroll

pauldambra commented:
> Looks great. The one opportunity I see is to automatically add the insight to a newly created dashboard. I noticed in your gif that you created the dashboard, it appended to the list, you clicked to add the insight. Can we collapse all that into the submit action for adding a new dashboard? For example:
> 
> 1. Click to create a new dashboard
> 2. Name the dashboard and submit
> 3. See the new dashboard added to the list and shown as `added`
> 
> Thoughts?

Yep, that would be nicer! The current flow is as implemented here... create the dashboard and then have to click to save it

I just had a quick run at implementing the nicer flow and there's a jumble of interacting logics that get very upset... Since this is behind a flag, ideally this could be a follow-up - I'll take your guide on whether we should block here though

clarkus commented:
Not a blocker, but definitely something we should track for improvement. Are the logic constraints mostly tied to components we're reusing or is there anything else I can keep in mind when working on this next time?

pauldambra committed:
key insightLogic on insight id _and_ filters_hash

pauldambra committed:
key insightLogic on insight id _and_ filters_hash

pauldambra committed:
Revert "key insightLogic on insight id _and_ filters_hash"

This reverts commit 7c8d0831952b4e0a51369f2b37c69d8e2da228b9.

pauldambra committed:
Revert "key insightLogic on insight id _and_ filters_hash"

This reverts commit 8c82dab5b50b49aa1935f1f14f91b3f4db90aea2.

pauldambra committed:
Merge branch 'master' into multiple_filter_hashes

pauldambra committed:
merge from multiple_filters_hashes

pauldambra committed:
use lemon divider

pauldambra committed:
use lemon modal

pauldambra committed:
Merge branch 'master' into multiple_filter_hashes

pauldambra commented:
> Are the logic constraints mostly tied to components we're reusing or is there anything else I can keep in mind when working on this next time?

@clarkus it's the interactions of kea logics... I'd really like to work on this, not just because it's a better experience but to grab @mariusandra and level up my knowledge. It's definitely fixable!

pauldambra committed:
add snapshot that shows we are selecting the wrong tile in the test

pauldambra committed:
remove snapshots that change per run after using them to confirm sql is now correct

pauldambra committed:
ensure test will fail if a dasboard tile does not match

pauldambra committed:
don't overlap filters in test

mariusandra commented on frontend/src/lib/components/SaveToDashboard/AddToDashboard.scss:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    .ant-modal-header {
+        padding: 0 0 8px;
+        border-bottom-width: 0;
+    }
+
+    .ant-modal-footer {
+        padding: 8px 0 0;
+        border-top-width: 0;
+        display: block;
+    }
+
+    .ant-modal-body {
+        padding: 0 0 8px;
+
+        .list-wrapper {
+            min-height: 220px;
+            margin-top: 8px;
+        }
+
+        .existing-links-info {
+            color: var(--muted-alt);
+            margin-top: 8px;
+
+            strong {
+                color: var(--text-default);
+            }
+        }
+    }
+
+    .ant-modal-title {
+        font-weight: 600;
+        font-size: 11px;
+        line-height: 24px;
+        letter-spacing: 0.5px;
+        text-transform: uppercase;
+        color: var(--text-muted);
+    }
+}
```
Going to tag @Twixes here for lemon input.

mariusandra commented on frontend/src/lib/components/SaveToDashboard/SaveToDashboardModal.tsx:
```diff
@@ -4,21 +4,143 @@ import { Select, Modal } from 'antd'
+    const { insightLoading } = useValues(insightLogic)
+    const { updateInsight } = useActions(insightLogic)
+
+    async function addToDashboard(dashboardToAdd: number): Promise<void> {
+        updateInsight({ ...insight, dashboards: [...(insight.dashboards || []), dashboardToAdd] }, () => {
+            reportSavedInsightToDashboard()
+            lemonToast.success('Insight added to dashboard', {
+                button: {
+                    label: 'View dashboard',
+                    action: () => router.actions.push(urls.dashboard(dashboardToAdd)),
+                },
+            })
+        })
+    }
+
+    async function removeFromDashboard(dashboardToRemove: number): Promise<void> {
+        updateInsight(
+            {
+                ...insight,
+                dashboards: (insight.dashboards || []).filter((d) => d !== dashboardToRemove),
+            },
+            () => {
+                reportRemovedInsightFromDashboard()
+                lemonToast.success('Insight removed from dashboard')
+            }
+        )
+    }
```
Can these functions be moved to `saveToDashboardModalLogic`? Ideally we want to move as much data manipulation complexity into kea as possible.

pauldambra replied:
Hmm, that's what the logics got upset about when I tried to have the new dashboard auto selected 🤔

mariusandra commented on frontend/src/lib/components/icons.stories.tsx:
```diff
@@ -59,3 +60,55 @@ export function Icons(): JSX.Element {
 }
+
+export function IconWithCountBubble(): JSX.Element {
+    const betweenOneAndTwenty = Math.floor(Math.random() * (20 - 1 + 1)) + 1
```
Can you remove the randomness?

Theoretically this breaks the component's purity, which React assumes we keep. Now every render renders something different, and it's not guaranteed when anything will be rendered.

Plus it'll be another thing to fix, once we introduce visual regression testing.

mariusandra reviewed (commented):
Looks really good! 👍  Some comments inline and below to get it over the line. Will also tag @Twixes to feedback the lemons.

1. Could we make this popup longer if you have a huge screen? It's weird to scroll for dashboards when there's all that space available.

![image](https://user-images.githubusercontent.com/53387/165720853-bedbaa0b-a3e6-45e1-9749-e81c04ed9e4e.png)

2. Can the number be capped at 99 max? There will be someone who will add an insight to 101 dashboards. Just `Math.max` will do, nothing special needed here. If there's space, "99+" with a smaller font would be even better.
![image](https://user-images.githubusercontent.com/53387/165721451-afeee66c-1c71-4fc4-b56b-5b83b0625375.png)

pauldambra committed:
Merge branch 'master' into multiple_filter_hashes

pauldambra committed:
Merge branch 'multiple_filter_hashes' into add_multi_dash_ui

yakkomajuri committed:
chore: fix plugin server tests (#9571)

Twixes commented on frontend/src/lib/components/icons.tsx:
```diff
@@ -1,7 +1,21 @@
+
+export function IconWithCount({ count, showZero, children }: PropsWithChildren<IconWithCountProps>): JSX.Element {
+    return (
+        <div style={{ position: 'relative' }}>
```
The sizing of this seems a bit off: it's 24x25px in "Add to dashboard" and in Storybook it looks like weirdly like this:

<img width="47" alt="Screen Shot 2022-04-28 at 13 59 51" src="https://user-images.githubusercontent.com/4550621/165747698-e9477bc7-232b-4bfc-9fef-35ea80a9340c.png">

`IconWithCount` should be exactly the same size as the underlying icon (so 24x24px by default, or 1em by 1em).

Twixes commented on frontend/src/lib/components/icons.scss:
```diff
@@ -22,3 +21,19 @@
+    right: -0.3rem;
+    color: var(--white);
+    background: var(--primary);
+    font-size: 0.6rem;
```
These values are a bit off in px terms, as 0.6rem is typically 9.6px (unless user changes root font size from 16px). They should generally be multiples of 1/16.
```suggestion
    font-size: 0.5625rem;
```

Twixes commented on frontend/src/lib/components/icons.scss:
```diff
@@ -22,3 +21,19 @@
+    width: 16px;
+    height: 16px;
+    line-height: 16px;
+    top: -0.3rem;
+    right: -0.3rem;
```
The count is a bit further from the center of the icon in the design as the icon is more easily recognizable then.
```suggestion
    top: -0.375rem;
    right: -0.375rem;
```

Twixes reviewed (changes requested):
Love the feature! Left some lemon (and general code+UX) input

pauldambra committed:
chore: remove rolled out flags for history log (#9530)

* remove feature flag history flag

* remove person history feature flag

* fix activity log story

* remove unnecessary button

Twixes committed:
feat(funnels): Connect new funnel bars to persons modal (#9546)

* Ensure "Relative to previous step" is correct

* Fix sizing

* Connect persons modal to funnel bars

* Format numbers

* Format percentages

* Update utils.test.ts

* Fix coloring

* Fix persons modal

* Fix which step gets opened

* Keep `openPersonsModalForStep` in `StepLegend`

pauldambra committed:
Update frontend/src/lib/components/icons.scss

Co-authored-by: Michael Matloka <dev@twixes.com>

pauldambra committed:
Update frontend/src/lib/components/icons.scss

Co-authored-by: Michael Matloka <dev@twixes.com>

pauldambra committed:
show 9+ if over 9 in IconWithCount

pauldambra committed:
move to lemonmodal properly

pauldambra committed:
lemonify the modal

pauldambra committed:
remove unused class

pauldambra committed:
extract dashboard relation row

pauldambra committed:
pluralize dashboards

pauldambra committed:
icon with count size

pauldambra commented:
I think I've addressed all of the feedback except the move of the functions into the logic. 

I'd like to get this merged (since it's behind a flag) so I can test the many-to-many insights in prod properly and handle the location of the functions as a follow-up

Thoughts @mariusandra @Twixes

pauldambra committed:
move add and remove dashboard methods into the logic

pauldambra committed:
fix more things

mariusandra commented on frontend/src/lib/utils.test.ts:
```diff
@@ -189,7 +189,7 @@ describe('pluralize()', () => {
     })
     it('handles plural cases', () => {
-        expect(pluralize(28321, 'member')).toEqual('28321 members')
+        expect(pluralize(28321, 'member')).toEqual('28 321 members')
```
So we have 28 321-member units?

mariusandra reviewed (changes requested):
The dashboard part looks good, but what is all the other stuff in this PR, regarding removing a feature flag, fixing the person modal, updating a kafka queue test, etc? 😕

pauldambra committed:
Merge branch 'master' into add_multi_dash_ui

pauldambra commented:
> The dashboard part looks good, but what is all the other stuff in this PR,

@mariusandra and there was me saying I was enjoying stacked PRs 🤦 

needed to merge from master locally... just the changes from this PR showing now

pauldambra committed:
Update frontend/src/lib/components/SaveToDashboard/SaveToDashboardModal.tsx

Co-authored-by: Marius Andra <marius.andra@gmail.com>

mariusandra commented on frontend/src/lib/components/SaveToDashboard/SaveToDashboardModal.tsx:
```diff
@@ -1,24 +1,151 @@
+    })
+
+    const { searchQuery, currentDashboards, orderedDashboards, scrollIndex } = useValues(logic)
+    const { setSearchQuery } = useActions(logic)
+    const { addNewDashboard } = useActions(logic)
```
```suggestion
    const { setSearchQuery, addNewDashboard } = useActions(logic)
```

mariusandra reviewed (approved):
Still not sure why these buttons aren't the same high, the rest LGTM:

![image](https://user-images.githubusercontent.com/53387/165925704-35aeed64-af5a-4b22-84d0-3612bb923db7.png)

pauldambra committed:
quick styling workaround

pauldambra committed:
Update frontend/src/lib/components/SaveToDashboard/SaveToDashboardModal.tsx

Co-authored-by: Marius Andra <marius.andra@gmail.com>

pauldambra committed:
Update frontend/src/lib/components/SaveToDashboard/SaveToDashboardModal.tsx

Co-authored-by: Marius Andra <marius.andra@gmail.com>
